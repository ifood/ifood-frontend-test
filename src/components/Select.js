import { useState, useEffect } from 'react'
import { array, func, string } from 'prop-types'
import styled from 'styled-components'

const Select = ({ id, onChange, options, text }) => {
  const [valueSelected, setValueSelected] = useState('')

  useEffect(() => {
    if (valueSelected) {
      onChange(id, valueSelected)
    }
  }, [id, onChange, valueSelected])

  return (
    <SelectStyle
      className="wrapper-filter-input wrapper-filter-select"
      valueSelected={valueSelected}
    >
      {options && options.length ? (
        <>
          <select
            className="filter-input filter-select"
            onChange={(event) => setValueSelected(event.target.value)}
          >
            <option />
            {options.map((option) => (
              <option key={option.value}>{option.name}</option>
            ))}
          </select>
          <label id="label-select">{text}</label>
        </>
      ) : null}
    </SelectStyle>
  )
}

const valueActive = `
  select.filter-select {
    color: white;
  }

  #label-select {
    color: white;
    top: -10px;
    z-index: 1;
  }
`

const SelectStyle = styled.div`
  .filter-select {
    cursor: pointer;
    -webkit-appearance: none;
    -moz-appearance: none;

    &:focus + label {
      color: white;
      top: -10px;
      z-index: 1;
    }
  }

  ${({ valueSelected }) => (valueSelected ? valueActive : '')}
`

Select.propTypes = {
  id: string.isRequired,
  onChange: func.isRequired,
  options: array.isRequired,
  text: string.isRequired
}

export default Select
