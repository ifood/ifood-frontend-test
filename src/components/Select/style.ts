import styled from 'styled-components';

export const SelectWrapper = styled.div`
	margin-bottom: 3px;
	@media (min-width: 800px) {
		margin-right: 5px;
	}
`;

export const Select = styled.select`
	width: 150px;
	height: 30px;
	border-radius: 6px;
	border: 1px solid #8b8b8b;
	outline: 0;
	font-size: 16px;

	:focus {
		border: 1px solid #ea1d2c;
	}

	@media (min-width: 900px) {
		width: 170px;
	}
`;