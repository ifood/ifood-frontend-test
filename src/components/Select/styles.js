import styled from 'styled-components'

export const SelectWrapper = styled.div`
  > select {
    font-size: 14px;
    width: 100%;
    height: 32px;
    background: none;
    border: none;
    outline: none;
    border-bottom: 2px solid;

    &:focus {
      border-bottom-color: ${(props) => props.theme.blue};
    }

    > options {
      margin-top: 8px;
    }
  }
`
