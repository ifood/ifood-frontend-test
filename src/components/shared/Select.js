import React from 'react';
import shortid from 'shortid';

const SingleOption = ({ value = '', name = '' }) => {
  return <option value={value}>{name}</option>
}

const Options = ({ values = [] }) => {
  return values.map(v =>
    <SingleOption
      key={shortid()}
      name={v.name}
      value={v.value}
    />
  );
}

const Select = ({
  selectedValue = '',
  name = '',
  values = [],
  initialValue = 'Selecione...',
  onChange = () => {}
}) => {
  return (
    <select name={name} className="form-control" onChange={onChange} value={selectedValue}>
      <SingleOption value="" name={initialValue} />
      <Options values={values} />
    </select>
  );
}

export default Select;