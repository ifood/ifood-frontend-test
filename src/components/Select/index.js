import React, { useContext } from "react";
import PropTypes from "prop-types";
import { FilterDispatchContext } from "../../context/Filter";

import { Container, SelectField, Label } from "./styles";

const Select = ({ field, value }) => {
  const dispatch = useContext(FilterDispatchContext);
  return (
    <Container>
      {field && (
        <div>
          <Label htmlFor={field.id}>{field.name}: </Label>
          <SelectField
            id={field.id}
            onChange={(event) =>
              dispatch({
                type: "UPDATE_SELECTED_FILTER",
                payload: { [field.id]: event.target.value },
              })
            }
            value={value}
          >
            <option value="">Selecione</option>
            {field.values.map((filter) => (
              <option value={filter.value} key={filter.value}>
                {filter.name}
              </option>
            ))}
          </SelectField>
        </div>
      )}
    </Container>
  );
};

Select.propTypes = {
  field: PropTypes.object,
  value: PropTypes.string,
};

export default Select;
