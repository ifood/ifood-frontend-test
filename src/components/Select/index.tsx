import React, { InputHTMLAttributes } from 'react';
import { StyledInputGroup } from './styles';

interface InputProps extends InputHTMLAttributes<HTMLSelectElement> {
  label: string;
  options: any[];
}

const Select = ({ label, options, ...props }: InputProps) => (
  <StyledInputGroup>
    <label htmlFor={props.id}>
      {label}
      {props.required ? '*' : ''}
    </label>
    <select {...props}>
      {options.map(option => (
        <option key={option.value} value={option.value}>
          {option.name}
        </option>
      ))}
    </select>
  </StyledInputGroup>
);

export default Select;
