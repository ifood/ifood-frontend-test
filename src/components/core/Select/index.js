import React, { useState } from "react";
import PropTypes from "prop-types";
import { useFormContext } from "react-hook-form";

import * as S from "./styles";

const Select = ({ register, optionId, values }) => {
  const [isOpen, setIsOpen] = useState(false);
  const { setValue } = useFormContext();

  return (
    <>
      <S.FieldSelect
        ref={register}
        id={optionId}
        name={optionId}
        type="text"
        onClick={() => setIsOpen(true)}
        readOnly
      />
      {isOpen ? (
        <>
          <S.Close onClick={() => setIsOpen(false)}>x</S.Close>
          <S.Options>
            {values?.map(({ name, value }) => (
              <S.Option
                key={value}
                onClick={() => {
                  setValue(optionId, value);
                  setIsOpen(false);
                }}
              >
                {name}
              </S.Option>
            ))}
          </S.Options>
        </>
      ) : (
        <S.Arrow onClick={() => setIsOpen(true)} />
      )}
    </>
  );
};

Select.propTypes = {
  register: PropTypes.func.isRequired,
  optionId: PropTypes.string.isRequired,
  values: PropTypes.array.isRequired,
};

export default Select;
