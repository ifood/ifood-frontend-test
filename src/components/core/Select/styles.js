import styled from "styled-components";

import { Field } from "components/core/Filter/styles";

export const FieldSelect = styled(Field)`
  position: relative;
`;

export const Options = styled.ul`
  position: absolute;
  top: 3.8rem;
  width: 100%;
  max-height: 16rem;
  overflow: scroll;
  z-index: 3;
  background: var(--black-rbg);
`;

export const Option = styled.li`
  padding: 1rem;
  list-style: none;
  font-size: 1.4rem;
  cursor: pointer;
  transition-property: background-color;
  transition-duration: 0.7s;
  transition-timing-function: cubic-bezier(0.5, 0, 0, 1);
  color: var(--gray);

  &:hover {
    background: var(--black-light);
  }
`;

export const Close = styled.span`
  position: absolute;
  top: 0.4rem;
  right: 0;
  font-weight: bold;
  font-size: 1.5rem;
  cursor: pointer;
  color: var(--white);
`;

export const Arrow = styled.span`
  position: absolute;
  top: 1.2rem;
  right: 0;
  cursor: pointer;
  border-left: 7px solid transparent;
  border-right: 7px solid transparent;
  border-top: 7px solid var(--white);
`;
