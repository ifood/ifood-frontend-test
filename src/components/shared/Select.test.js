import React from 'react'
// import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Select from './Select';

const defaultProps = {
  selectedValue: '',
  name: 'field-name',
  values: [
    { name: 'First item', value: '1' },
    { name: 'Second item', value: '2' },
    { name: 'Third item', value: '3' }
  ],
  initialValue: 'Selecione...',
  onChange: () => {}
}

describe('<Select />', () => {

  it('Should render correctly', () => {
    const elem = renderer
      .create(<Select { ...defaultProps } />)
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

});

