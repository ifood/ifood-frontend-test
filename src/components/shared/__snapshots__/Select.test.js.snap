// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Should render correctly 1`] = `
<select
  className="form-control"
  name="field-name"
  onChange={[Function]}
  value=""
>
  <option
    value=""
  >
    Selecione...
  </option>
  <option
    value="1"
  >
    First item
  </option>
  <option
    value="2"
  >
    Second item
  </option>
  <option
    value="3"
  >
    Third item
  </option>
</select>
`;
